### PR TITLE
Initial support for Windows

### PR DESCRIPTION
**Pull request by azchohfi**
#489

### CODE REVIEW

**Comment by azchohfi** [comment]
I still need to add a Windows workflow.

**Comment by Gustl22** [comment]
I tried your branch and after some additional changes I got it working 🎉 :
- I had to install additional Visual Studio components:
  - C++ ATL for latest v143 build tools (x86 & x64)
  - C++/WinRT
- I had to [link the windowsApp lib](https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/faq#why-is-the-linker-giving-me-a--lnk2019--unresolved-external-symbol--error-) `#pragma comment(lib, "windowsapp")` in every .cpp file in `packages/audioplayers/windows/`

If I play two different samples multiple times alternately I get the error:
```
[ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: FileSystemException: Cannot open file, path = 'C:\Users\Gustl\AppData\Local\Temp/assets/audio/AirHornShip.mp3' (OS Error: The process cannot access the file because it is being used by another process.

, errno = 32)
#0      _File.open.<anonymous closure> (dart:io/file_impl.dart:356:9)
#1      _rootRunUnary (dart:async/zone.dart:1434:47)
#2      _CustomZone.runUnary (dart:async/zone.dart:1335:19)
<asynchronous suspension>
#3      AudioCache.fetchToMemory (package:audioplayers/src/audio_cache.dart:94:5)
<asynchronous suspension>
#4      AudioCache.load (package:audioplayers/src/audio_cache.dart:115:31)
<asynchronous suspension>
#5      AudioCache.play (package:audioplayers/src/audio_cache.dart:159:17)
<asynchronous suspension>
```

Thank you for your great work!

**Comment by azchohfi** [comment]
@Gustl22 can you share your code that simulates this issue? Also, is this happening only on the Windows implementation?

**Comment by Gustl22** [comment]
Seems like you are supposed to have only one `AudioCache` which is also recommended by the library in order to save loading times. 

[Here](https://github.com/bluefireteam/audioplayers/blob/main/packages/audioplayers/doc/audio_cache.md) audio cache is described to have its own cache for each `AudioCache`-instance. So when having multiple AudioCaches while loading also the same file, one AudioCache is not able to override the file in the cache directory, which is still used by the other AudioCache. 

But the question remains, why the reference is held, even when the AudioCache and player is closed. But one should always use only one AudioCache, as this case can also happen, playing the same file twice simultaneously.
So would not consider this as flaw of the windows implementation, also if it works on other Platforms.

**Comment by azchohfi** [comment]
I believe I have found the issue:
https://github.com/dart-lang/sdk/issues/31557
I will investigate more on Monday, but I believe that is it. Not sure if there is a simple workaround.

**Comment by luanpotter** [comment]
Thank you so much for this effort!
My goal is to get [the federation PR](https://github.com/bluefireteam/audioplayers/pull/1051) merged first, and then we can rebase this to add support for Windows and Linux into their own packages.
I am just wrapping up some simplifications on the platform interface to make each new platform easier to add, and should be ready to merge soon!

**Comment by azchohfi** [comment]
Awesome! Waiting on #1051 to get merged then.

**Comment by fzyzcjy** [comment]
Looking forward to this - super useful! Especially because windows is stable now and I am adding windows support for my app.

**Comment by fzyzcjy** [comment]
#1051 seems to be merged 🎉

**Comment by azchohfi** [comment]
This should be working now, and rebased on main.

File: packages/audioplayers/pubspec.yaml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 name: audioplayers
 description: A Flutter plugin to play multiple audio files simultaneously
-version: 0.20.1
+version: 0.21.0
```
Comment (luanpotter) [inline packages/audioplayers/pubspec.yaml:3]: we can leave this as is, I will bump everything later (probably will add melos for repo-wide management)
```suggestion
version: 0.20.1
```

File: packages/audioplayers_windows/CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,298 @@
+# Changelog
```
Comment (luanpotter) [inline packages/audioplayers_windows/CHANGELOG.md:1]: we should probably make these links like the other files, but I will change that on a followup

File: packages/audioplayers_windows/windows/Logger.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#include <string>
+
+#include <iostream>
+
+enum class LogLevel
+{
```
Comment (luanpotter) [inline packages/audioplayers_windows/windows/Logger.h:6]: unless there is some enum/cpp specific convention I don't know, I think this should follow the rule used by other classes (like Logger) or blocks right?
```suggestion
enum class LogLevel {
```

Comment (luanpotter) [reply packages/audioplayers_windows/windows/Logger.h:6]: actually I am seeing a mix of stuff here. let's merge as is and I will study c++ best practices/add an auto linter later

File: packages/audioplayers_windows/windows/audioplayers_windows_plugin.cpp [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    result->Success(EncodableValue(1));
+  } else if (method_call.method_name().compare("seek") == 0) {
+    auto position = GetArgument<int>("position", args, (int)(player->GetPosition() / 10000));
+    player->SeekTo(position * 10000);
+    result->Success(EncodableValue(1));
+  } else if (method_call.method_name().compare("setSourceUrl") == 0) {
+    auto url = GetArgument<std::string>("url", args, std::string());
+
+    if (url.empty()) {
+        Logger::Error("Null URL received on setUrl");
```
Comment (luanpotter) [inline packages/audioplayers_windows/windows/audioplayers_windows_plugin.cpp:190]:
```suggestion
        Logger::Error("Null URL received on setSourceUrl");
```

File: packages/audioplayers_windows/windows/audioplayers_windows_plugin.cpp [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+        result->Success(EncodableValue(0));
+        return;
+    }
+
+    try {
+      player->SetUrl(url);
```
Comment (luanpotter) [inline packages/audioplayers_windows/windows/audioplayers_windows_plugin.cpp:196]:
```suggestion
      player->setSourceUrl(url);
```

File: packages/audioplayers_windows/windows/audioplayers_windows_plugin.cpp [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+      if (seekTimeMillis != -1) {
+        player->SeekTo(seekTimeMillis * 10000);
+      }
+      player->Play();
+      result->Success(EncodableValue(1));
+    } catch(...) {
+      Logger::Error("Error playing Url:'" + url + "'.");
+      result->Success(EncodableValue(0));
+    }
+  } else if (method_call.method_name().compare("pause") == 0) {
```
Comment (luanpotter) [inline packages/audioplayers_windows/windows/audioplayers_windows_plugin.cpp:168]:
```suggestion
  if (method_call.method_name().compare("pause") == 0) {
```

File: packages/audioplayers_windows/windows/audio_player.cpp [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    auto onTimeUpdateCB = std::bind(&AudioPlayer::OnTimeUpdate, this);
+    auto onSeekCompletedCB = std::bind(&AudioPlayer::OnSeekCompleted, this);
+
+    // Create and initialize the MediaEngineWrapper which manages media playback
+    m_mediaEngineWrapper = winrt::make_self<media::MediaEngineWrapper>(nullptr, onError, onBufferingStateChanged, onPlaybackEndedCB, onTimeUpdateCB, onSeekCompletedCB);
+
+    m_mediaEngineWrapper->Initialize();
+}
+
+void AudioPlayer::SetUrl(std::string url) {
```
Comment (luanpotter) [inline packages/audioplayers_windows/windows/audio_player.cpp:34]:
```suggestion
void AudioPlayer::SetSourceUrl(std::string url) {
```

File: packages/audioplayers_windows/windows/audioplayers_windows_plugin.cpp [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+  }
+}
+
+AudioPlayer* AudioplayersWindowsPlugin::GetPlayer(std::string playerId, std::string mode) {
+  auto searchPlayer = audioPlayers.find(playerId);
+  if(searchPlayer != audioPlayers.end()) {
+    return searchPlayer->second.get();
+  } else {
+    auto player = std::make_unique<AudioPlayer>(playerId, channel.get());
+    //player->Init();
```
Comment (luanpotter) [inline packages/audioplayers_windows/windows/audioplayers_windows_plugin.cpp:235]: I'm assuming this isn't needed?
```suggestion
```

**Review by luanpotter — APPROVED** [review]
Awesome work @azchohfi ! Thank you so much for this invaluable contribution and for the patience with our reorganization of the repo to support federation.
Contributions like this make me happy to be a part of this community :)

LGTM!

File: packages/audioplayers_windows/windows/audio_player.h [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    void SetPlaybackSpeed(double playbackSpeed);
+    void Play();
+    void Pause();
+    void Resume();
+    bool GetLooping();
+    int64_t GetPosition();
+    int64_t GetDuration();
+    void SeekTo(int64_t seek);
+
+    void SetUrl(std::string url);
```
Comment (luanpotter) [inline packages/audioplayers_windows/windows/audio_player.h:73]:
```suggestion
    void SetSourceUrl(std::string url);
```

**Pull request merged**